libyara/modules/pe/pe.c

```diff
@@ -171,91 +172,99 @@ static void pe_parse_rich_signature(
   if (yr_le16toh(mz_header->e_magic) != IMAGE_DOS_SIGNATURE)
     return;
 
-  if (yr_le32toh(mz_header->e_lfanew) < 0)
+  // To find the Rich marker we start at the NT header and work backwards, so
+  // make sure we have at least enough data to get to the NT header.
+  nthdr_offset = yr_le32toh(mz_header->e_lfanew);
+  if (nthdr_offset > pe->data_size + sizeof(uint32_t) || nthdr_offset < 4)
     return;
 
-  headers_size = yr_le32toh(mz_header->e_lfanew) + \
-                 sizeof(pe_header->Signature) + \
-                 sizeof(IMAGE_FILE_HEADER);
+  // Most files have the Rich header at offset 0x80, but that is not always
+  // true. 582ce3eea9c97d5e89f7d83953a6d518b16770e635a19a456c0225449c6967a4 is
+  // one sample which has a Rich header starting at offset 0x200. To properly
+  // find the Rich header we need to start at the NT header and work backwards.
+  p = (DWORD*)(pe->data + nthdr_offset - 4);
+  while (p > (DWORD*)(pe->data + sizeof(IMAGE_DOS_HEADER)))
```

Comment:
I think that doing everything in the same loop makes the code more complicated than necessary. Why not searching for the Rich header in one loop, and once you find, continue in another loop that computes the signature starting at where the previous loop left? 

This way you avoid the `key == 0` part in `*p == RICH_RICH && key == 0` and the `if (key != 0)` statement and the code is easier to follow.

Reply:
I thought about doing that, will update it tonight.